crates/pg_workspace_new/src/workspace/server.rs

```diff
@@ -291,6 +300,50 @@ impl Workspace for WorkspaceServer {
     fn is_path_ignored(&self, params: IsPathIgnoredParams) -> Result<bool, WorkspaceError> {
         Ok(self.is_ignored(params.pglsp_path.as_path()))
     }
+
+    fn pull_diagnostics(
+        &self,
+        params: super::PullDiagnosticsParams,
+    ) -> Result<super::PullDiagnosticsResult, WorkspaceError> {
+        // get all statements form the requested document and pull diagnostics out of every
+        // sourcece
+        let doc = self
+            .documents
+            .get(&params.path)
+            .ok_or(WorkspaceError::not_found())?;
+
+        let diagnostics: Vec<SDiagnostic> = doc
+            .statement_refs_with_ranges()
+            .iter()
+            .flat_map(|(stmt, r)| {
+                let mut stmt_diagnostics = vec![];
+
+                stmt_diagnostics.extend(self.pg_query.pull_diagnostics(stmt));
+
+                stmt_diagnostics
+                    .into_iter()
+                    .map(|d| {
+                        SDiagnostic::new(
+                            d.with_file_path(params.path.as_path().display().to_string())
+                                .with_file_span(r),
+                        )
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .collect();
+
+        let errors = diagnostics
+            .iter()
+            .filter(|d| d.severity() == Severity::Error)
```

Comment:
There's also `Severity::Fatal` which would not be included here, should we maybe check `d.severity() >= Severity::Error`?